proposals/0022-multi-stake.md

```diff
@@ -0,0 +1,178 @@
+less than the minimum delegation, but they won't earn rewards.
+
+Unfortunately, small stakes can still bloat the stakes cache from a memory level.
+
+* Don't include small stakes in the cache, but track them in transactions
+
+In this solution, on every transaction, if the runtime sees a small stake, it
+includes its delegation amount in the validator's voting power, but doesn't
+add its pubkey and `Delegation` to the cache.
+
+The runtime tracks the pre and post state of every account, and for a successful
+transaction, subtracts the old small stake delegation amount, and adds the new small
+stake delegation amount.
+
+During rewards payout, small stakes are completely omitted, since they are not
+present in the cache.
+
+This is incredibly brittle, however, and introduces more overhead in the runtime.
+The bank must hold onto pre-states for all transactions to debit the stakes cache. And,
+if called from the outside, all `store_accounts` functions on the bank could easily
+invalidate the stakes cache.
+
+For example, if you store a small stake directly, what does the cache do? What
+was the pre-state of that small stake? There's no way to know.
+
+## Detailed Design
+
+Rather than hacking the validator, let's change the stake program.
+
+With a new "multi stake" account, carrying *two* more `Delegation`s, the stake
+program becomes much more flexible.
+
+During initial delegation, the "multi stake" account must have enough to cover
+the minimum stake delegation amount. The two other `Delegation` instances may be
+used for small stake movements.
+
+Here are some example uses:
+
+### Add or remove small stake
+
+To add a small stake to your account, transfer the lamports, call `delegate` once
+more, and now your stake account has two delegations: the initial one, and the
+newly activating one.
+
+Once the new `Delegation` is active, a permissionless `update` instruction follows
+the existing `merge` logic to consolidate the two `Delegation`s into one.
+
+The same roughly applies for removing a small stake. You `deactivate` a portion
+of it, which adds another delegation to your stake account: the initial one, and
+the newly deactivating one.
+
+### Redelegate
+
+With another `Delegation` instance, you can redelegate from one validator to another
+within the same account.
+
+It works the same as the existing "redelegate" instruction, except the lamports
+all stay in the same account, and the second `Delegation` covers the redelegation.
+
+The permissionless `update` instruction clears out the first one once it's inactive.
+
+### Single validator stake pool
+
+A single-validator stake pool program can manage small stakes with 100% efficiency
+using "multi stakes".
+
+Since there's a total of *three* `Delegation` instances, small stakes can be added
+and removed at the same time. One instance covers the main amount, another covers
+new activations, and the last one covers deactivations.
```

Comment:
the third `Delegation` improves capital efficiency over independent activation and deactivation transient accounts by being able to share active reserve pool. correct?

Reply:
to be honest, I was focused on the ability to activate and deactivate small amounts at the same time, but this will be another great side effect!

Reply:
Actually, after more thought about this, you made me realize that we only need two `Delegation` instances total to cover all possible movements. This will also simplify the implementation a lot.